src/feature.js

```diff
@@ -183,27 +201,43 @@ export default class Feature {
    * If it has multiple values, those values will be concatenated with a default separator and
    * returned in a single string. Values composing this string are sorted according
    * to each value's sort order.
+   * NOTE: If object contains a single value and it is a number, it will be converted to a string.
    * @return {string} A single value string.
    */
   get value () {
     return this.values.join(this.constructor.joinSeparator)
   }
 
+  /**
+   * Returns a feature value, if Feature object contains a single value. If no value is stored,
+   * returns `undefined`. If feature has more than one value, throws an error.
+   * This method allows to avoid conversion of a value to the string type as is the case
+   * with other methods.
+   *
+   * @returns {undefined|*} - A single value in a format in which it is stored or `undefined`
+   *          if feature has no value.
+   */
+  get singleValue () {
+    if (this._data.length === 0) return undefined
+    if (this._data.length > 1) throw new Error('More than one value stored')

```

Comment:
I didn't understand, why it would return a Error.
I believe a feature could be multy-valued.

Reply:
Also I guess you define a constant with Error text for this case?

Reply:
so the intent of this method is to request and confirm that a feature is single-valued, rather than retrieving a single value from a possibly multi-valued feature?

Reply:
This is a good catch on not using a constant there. I will fix that.

The problem I was trying to solve is that we did not have a method to retrieve a single value without converting to to string (that's the conversion that a `value` getter does).

We have `values` that returns one or several values without conversion but using it to get a single value would be not smooth as we have to do something like `feature.values[0]`.

So I decided to create a getter that will be used to retrieve a single value. It is supposed to be used for features that have a single value only (we have plenty of cases for that). And if someone would mistakenly use it for features that contain multiple values I thought it would make sense to throw an error to warn a user about that mistake. @irina060981, @balmas does it makes sense? Or shall we handle it differently?

Reply:
It's okay with me but I just wonder in what situation client code might use it, instead of just assuming every feature it retrieves *could be* multi-valued.

Reply:
I don't know if it's the right approach, but the idea was as following. Feature is a universal container that can hold both multiple values and a single value as well. In majority of cases it will probably hold a single value. Probably more than that, in some cases we know that some features would always hold a single value, as radical, for example. So wouldn't it be beneficial to separate those two into the separate worlds of their own as they may never overlap (i.e a radical feature would never ever hold multiple values)? 

Working with single feature values and reading the code that handles such features it might be beneficial:
* To understand by looking at the code that this particular feature would have a single value only and should be treated according (i.e. we should not loop over all possible values as there will be only one). I think it will help to understand the specifics of a particular feature better.
* To protect single-value features from being treated as multi-value ones and vice versa (i.e. throw an error when a multi-value feature is accessed as a single-value one and vice versa).

Feature object is very versatile, but versatility comes at a price and I feel sometimes it might be beneficial to limit this versatility. That might result in a code that will be easier to understand and is less error prone. Flexibility increases the complexity and sometimes more a limited but simpler to understand and easier to verify and control approach is all we need. It's like languages with no type system are infinitely more flexible yet a type system, while limiting the flexibility, might often result being more beneficial to the overall quality of the code. Sometimes we need the flexibility but sometimes it will be better to give it away to benefit in something else.

That was the general idea behind having a `singleValue` method. It's not a full-blown effort but a small step in that direction. But maybe is not the right direction, maybe it's not fully in line with where we are and where we are going, and maybe it's simply an overkill. Maybe using `values[0]` is all we need.

@balmas, @irina060981 what do you think? Do we need to have a `singleValue` method that will clearly state this feature will have as single value forever and ever and thinking otherwise would be a mistake or is it better to stay flexible and versatile and use `values[0]` for both single-value and multi-value features instead? In that case I would gladly remove this method in favor of a more versatile `values` approach.

Reply:
I agree generally with your statement about versatility vs complexity.   I don't think there is any harm in having a singleValue method.

Reply:
I believe the problem with Feature in understanding what is the value of some Feature.
Is it a string or an array in some case? and how to treat with it because we don't know what is the type of it. But by the meaning there is no difference - if it is `masculine` or `masculine femine`, for example, from the code perspective.

May be it is better to create:
  a) one method `value` with additional property - format, that could be single, concated string, array and may be some other?
  b) or methods by type - valueString, valueArray ?

This way it would be really easy to understand what is the type of the Feature property and how to handle with it, without mixing with the sense - how much values could have a feature at a time?

Reply:
> I believe the problem with Feature in understanding what is the value of some Feature.
> Is it a string or an array in some case? and how to treat with it because we don't know what is the type of it. But by the meaning there is no difference - if it is `masculine` or `masculine femine`, for example, from the code perspective.

Exactly that! But I would say sometimes knowing the list of allowed values a feature holds can be beneficial too. For example, when we compare a feature value with something it is important to know what it can be compared with.

On my opinion, this was not a problem when Alpheios worked with Latin and Greek where features were simpler and easier to understand. Mult-value features where more an exception than a norm. But it became much more complex with Chinese, where we not only need features to be multi-valued but also to carry some additional information as well, such as the `classifier` type (Tang, Mandarin, Cantonese) for the pronunciation feature.

So, ideally, it would be great to have something that, for each particular feature type, describe what value type will the feature hold, what values will be allowed, and how would it store some additional information, if necessary. Maybe something in a similar vein with PropTypes: https://github.com/facebook/prop-types (docs mentions it for React only but now it can work with any library type).

Reply:
We had started something along this lines with the typeFeatures structure in the language-model class, where the language model class defines the allowed values for a feature. How does that come into play here?

Reply:
FeatureType class is now deprecated as we moved its functionality to the Feature object. Feature object has a list of allowed values so that it can control what possible values the feature can have. If we add a prop saying how many values may this feature has and of what type those features can be maybe it will get us covered.

With all those implemented we can probably create all "feature types" for the language as constants of a Feature class within the language model. Each feature type, having all the knowledge of possible feature values, can serve as a generator of features of a certain type (using `createFeature(s)` method). Looking at the "feature type" constant we can understand what values the feature will hold.

Would that work? What do you think?

Reply:
I think so... will have to see the code :-)